src/main/java/org/jenkinsci/remoting/util/Timeout.java

```diff
@@ -0,0 +1,108 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.remoting.util;
+
+import hudson.remoting.VirtualChannel;
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Allows operations to be limited in execution time.
+ * For example, {@link VirtualChannel#call} could otherwise may hang forever.
+ * Use in a {@code try}-with-resources block.
+ * @author Jesse Glick
+ * @author Oleg Nenashev
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class Timeout implements AutoCloseable {
+
+    private static final Logger LOGGER = Logger.getLogger(Timeout.class.getName());
+
+    private static final Timeout NO_TIMEOUT = new Timeout(null);
+    
+    @CheckForNull
+    private final ScheduledFuture<?> task;
+
+    private Timeout(@CheckForNull ScheduledFuture<?> task) {
+        this.task = task;
+    }
+
+    @Override 
+    public void close() {
+        if (task != null) {
+            task.cancel(false);
+        }
+    }
+
+    /**
+     * Sets optional timeout.
+     * 
+     * Disabled timeout is a not recommended mode for legacy operations.
+     * A warning will be printed to the system log if it is defined.
+     * 
+     * @param timeout Timeout duration. If {@code null}, no timeout is set
+     * @return Timeout object
+     */
+    public static Timeout optLimit(final @CheckForNull Duration timeout) {
+        if (timeout != null) {
+            return limit(timeout);
+        } else {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Executing code block with disabled timeout. " +
+                        "It may hang indefinitely.", new IllegalStateException("Timeout duration is null"));
+            }
+            return NO_TIMEOUT;
+        }
+    }
+    
+    public static Timeout limit(final @Nonnull Duration timeout) {
+        return limit(timeout.toMillis(), TimeUnit.MILLISECONDS);
+    }
+    
+    public static Timeout limit(final long delay, final @Nonnull TimeUnit unit) {
+        final Thread thread = Thread.currentThread();
+        return new Timeout(Timer.get().schedule(new Runnable() {
+            @Override public void run() {
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    Throwable t = new Throwable();
+                    t.setStackTrace(thread.getStackTrace());
+                    LOGGER.log(Level.FINE, "Interrupting " + thread + " after " + delay + " " + unit, t);
+                }
+                thread.interrupt();
```

Comment:
I think this utility is inappropriate here.

I introduced it in Pipeline code (later elsewhere) because I was calling predefined APIs (`Channel.call`, ultimately) which offered no timeout (à la `Future.get(long, TimeUnit)`), and from reading the source code I knew that the implementation would _typically_ be inside an `Object.wait()` loop in `Request.call`, which I had no control over and which waited with no bound except a response, channel closing, or a thread interruption. So this was certainly a workaround.

But if you are implementing timeouts in `remoting` itself, there is no reason to resort to that. You can simply make `Request.call` not wait forever. It can `wait` for the configured timeout, and not use the `while` loop.